Prevent PointValues from returning null for ghost fields

javanna opened the pull request:
getPointValues may currently return null for unknown fields or fields that don't index points. It can happen that a field no longer has points for any document in a segment after delete+merge, which causes field info to think that the field is there and has points, yet when calling getPointValues null is returned.

With this change, we prevent getPointValues from returning null for ghost fields, it will instead return an empty instance of PointValues.

Disclaimer: this is an attempt around addressing potential issues with ghost fields when retrieving point values. It has to be said that most places that call getPointValues today have null checks, and don't consult field info upfront, hence they would not be exposed to the issue that FieldExistsQuery triggers, as field info says there are points and then the retrieved point values is null. One quick-fix would be to add a null check to FieldExistsQuery, though it feels like we should try to prevent this kind of situation from happening again in the future? getPointValues can still return null, but it shouldn't for ghost fields?

Relates to #11393

javanna committed:
Prevent PointValues from returning null for ghost fields

getPointValues may currently return null for unknown fields or fields that don't index points.
It can happen that a field no longer has points for any document in a segment after delete+merge, which
causes field info to think that the field is there and has points, yet when calling getPointValues null
is returned.

With this change, we prevent getPointValues from returning null for ghost fields, it will instead return an
empty instance of PointValues.

Relates to #11393

javanna commented on lucene/core/src/java/org/apache/lucene/index/PointValues.java:
```diff
@@ -467,4 +467,91 @@ public final long estimateDocCount(IntersectVisitor visitor) {
+      new PointTree() {
+        @Override
+        public PointTree clone() {
+          throw new UnsupportedOperationException();
```
there are not test failures caused by this but I have been wondering if this is acceptable.

javanna commented on lucene/core/src/java/org/apache/lucene/index/PointValues.java:
```diff
@@ -467,4 +467,91 @@ public final long estimateDocCount(IntersectVisitor visitor) {
+
+        @Override
+        public byte[] getMaxPackedValue() {
+          return null;
```
returning null here does not feel great, yet I did not find a better way that fits in the existing code. I tried the empty array but it required changes in how the return value is consumed.

jpountz commented on lucene/core/src/java/org/apache/lucene/index/PointValues.java:
```diff
@@ -467,4 +467,91 @@ public final long estimateDocCount(IntersectVisitor visitor) {
+        public void visitDocValues(IntersectVisitor visitor) {}
+      };
+
+  public static final PointValues EMPTY =
```
One issue that the other PR regarding ghost fields and terms highlighted is that we should tune this empty instance so that it returns correct values for field metadata such as numbers of dimensions. Otherwise any code that would be checking the number of dimensions via this empty instance could be retrieving wrong numbers. See e.g. `checkValidPointValues` in `PointRangeQuery`.

javanna committed:
adjust dimension counts in empty instance

javanna committed:
spotless

jpountz reviewed (changes requested):
This looks like the right path forward to me. Can you update javadocs of `LeafReader#getPointValues` to clarify when `null` may be returned, and maybe look for places where we would be able to remove some null checks, e.g. `scorerSupplier` in `RangeFieldQuery` or `PointInSetIncludingScoreQuery`?

javanna committed:
iter

javanna committed:
spotless

javanna commented:
I pushed an update. I have removed null checks from consumers that were already checking field info. In most cases we already check field info for compatibility, hence we can expect point values to never be null. Where we instead we only check if field info is null, we also need to add a check that point dimension count is > 0 to replace the point values null check.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java:
```diff
@@ -104,28 +104,28 @@ public abstract class NumericLeafComparator implements LeafFieldComparator {
-        }
+      FieldInfo info = context.reader().getFieldInfos().fieldInfo(field);
+      if (info == null || info.getPointDimensionCount() == 0) {
+        throw new IllegalStateException(
```
This doesn't look correct to me as it would throw an exception if the field doesn't exist, which should be supported. Previously it would only fail if the field would not exist AND the reader provides point values?

javanna replied:
yes this is wrong, I pushed a fix: the idea is that we don't need to check consistency between FieldInfo and getPointValues here, hence the first exception can go away. We check fieldinfo and we let that drive our decisions, and retrieve points based on what field info says.

jpountz commented:
I merged your other PR that adds a null check in FieldExistsQuery, we should now be able to remove this null check with this change?

javanna committed:
Merge branch 'main' into fix/ghost_fields_point_values

javanna committed:
remove field exists query null checks

javanna committed:
iter

javanna committed:
spotless

jpountz reviewed (approved):
This looks good to me. Can you add a CHANGES entry under 9.5?

jpountz commented:
Test failures suggest CheckIndex needs to have its expectations adjusted.

javanna committed:
update changelog

javanna committed:
add clarifying comment

javanna committed:
address check index

javanna committed:
Merge branch 'main' into fix/ghost_fields_point_values

javanna committed:
spotless

javanna committed:
rephrase comment

javanna committed:
Merge branch 'main' into fix/ghost_fields_point_values

javanna committed:
restored needed null check

javanna committed:
adjusted field info checks

We need to check the dimension count if we want to remove null checks for PointValues

javanna committed:
Handle situations where minPackedValue and maxPackedValue are null

When we encounter a ghost field, we would previously have thrown NPE in some places.
The PointValues instance is now non-null, but its packed values are null, so we need to make sure that if packed values are used, we prevent NPEs. One solution is to check the size of the point values: if 0, for sure the packed values will be null.

javanna committed:
tidy

javanna commented:
@jpountz would you have time to take another look at this please?

javanna commented:
Agreed @jpountz I think it was a good experiment to spend some time on, and I have also been thinking along the same lines, that the changes I ended up making were not solving the problem like we initially thought. No problem at all. I will open a follow-up PR to clarify the current behaviour and expectations.
